## Ομάδα 

Στοιχεία Μέλους 1: Σκουρτσή Δήμητρα-Άννα: el18044

Στοιχεία Μέλους 2: Στεργιόπουλος Ανστάσιος: el18605

# Τεχνητή Νοημοσύνη: Εργαστηριακή Άσκηση 2
---

Ο στόχος της εργασίας είναι η κατασκευή ενός συστήματος προτάσεων (Recommendation System) για ταινίες. Οι προτάσεις αυτές θα πηγάζουν τόσο από τα χαρακτηριστικά της ταινίας όσο και από ορισμένες αξιολογήσεις του κάθε χρήστη.

Στα δεδομένα της άσκησης περιλαμβάνονται ένα αρχείο με το όνομα movies_metadata.csv το όποιο περιέχει τα χαρακτηριστικά κάθε ταινίας όπως θέμα, σκηνοθέτης ηθοποιοί, λέξεις κλειδιά κ.α. από το imdb καθώς και τα αρχεία ratings.csv τα όποια περιέχουν πραγματικές αξιολογήσεις χρηστών, χωρισμένες σε train και σε test.


# Εκφώνηση
Στην παρούσα εργασία σας ζητείται να μελετήσετε και να υλοποιήσετε τα παρακάτω:<br>

### Μέρος 1
Στο μέρος 1 και 2 θα εργαστείτε μόνο με το αρχείο movies_metadata.csv, ενώ στο μέρος τρία θα δουλέψετε και με τα αρχεία των αξιολογήσεων.
#### Ερώτημα 1α

Αρχικά θα πρέπει, αφού μελετήσετε τη δομή και τα χαρακτηριστικά του movies_metadata.csv, να κατασκευάσετε μια βάση γνώσης για την Prolog η όποια ουσιαστικά θα αποτελεί τον κόσμο με τον όποιο θα εργαστείτε στην συνέχεια. Τα κατηγορήματα που θα δημιουργηθούν θα σας βοηθήσουν και στην κατασκευή του recommender και θα είναι της μορφής:

```
director(Movie, Director).
genre(Movie, Genre).
```
#### Ερώτημα 1β

Αφού δημιουργήσετε τον κόσμο του προβλήματος, στη συνέχεια καλείστε να δημιουργήσετε, σε Prolog, απλούς κανόνες οι οποίοι θα βρίσκουν όλες τις ταινίες με:
1.	Κοινό θέμα (κάποιες λέξεις σχετικά με το genre κοινές)
2.	Αρκετά κοινό θέμα (κάποιες λιγότερες λέξεις σχετικά με το genre κοινές π.χ. 3)
3.	Σχετικά κοινό θέμα (λίγες λέξεις σχετικές με το genre κοινές π.χ. 1)
4.	Κοινός σκηνοθέτης
5.	Ακριβώς ίδια πλοκή (κάποιες λέξεις κλειδιά της πλοκής κοινές)
6.	Σχετικά ίδια πλοκή (κάποιες λιγότερες λέξεις κλειδιά κοινές)
7.	Ίδιους τους βασικούς ηθοποιούς (και τους 3)
8.	Αρκετά ίδιους βασικούς ηθοποιούς (ορισμένους βασικούς ηθοποιούς κοινούς π.χ. 2)
9.	Σχετικά ίδιους ηθοποιούς (π.χ. 1 από τους 3)
10.	Ίδια γλώσσα
11.	Είναι έγχρωμες ή ασπρόμαυρες
12. Κοινό studio παραγωγής
13. Κοινή χώρα παραγωγής 
14. Ίδια δεκαετία 

Αξίζει να σημειωθεί ότι στα παραπάνω μπορείτε να προσθέσετε περισσότερα ερωτήματα ή να αλλάξετε την κλιμακωσιμότητα των queries (πέρα από το ίδιο, αρκετά ίδιο, σχετικά ίδιο) που θα κατασκευάσετε(π.χ. μια κλίμακα από 1 στα 5 όπου αυτό είναι δυνατόν), μιας και αυτά στην συνέχεια θα χρησιμοποιηθούν για την λειτουργία του recommender. Έτσι μπορείτε να προσθέσετε queries με τα όποια μπορεί να παράγονται καλύτερες συστάσεις  (το αρχείο movie_metadata.csv περιέχει πολλές πληροφορίες ακόμα για κάθε ταινία όπως έτος κυκλοφορίας, βαθμολογία στο imdb, facebοok_likes κ.α.). Το συγκεκριμένο μέρος εργασίας είναι προπαρασκευαστικό οπότε όσο καλύτερα και πλουσιότερα τα ερωτήματα που θα φτιάξετε σε αυτό το μέρος τόσο καλύτερη θα είναι η απόδοσή των συστημάτων συστάσεων των επόμενων ερωτημάτων.

Περισσότερες πληροφορίες για το dataset μπορείτε να διαβάσετε σε αυτό το [Link](https://www.kaggle.com/georgefila/movies-metadata):
### Μέρος 2: Recommendation System
Στο σημείο αυτό καλείστε με βάση αυτά που κάνατε στο μέρος 1 να κατασκευάσετε queries τα όποια θα σας επιστρέφουν παρόμοιες (σε χαρακτηριστικά) ταινίες. Τα ερωτήματα αυτά θα είναι κλιμακούμενα, δηλαδή θα υπάρχουν ερωτήματα που επιστρέφουν αρκετά κοινές ταινίες αλλά και που επιστρέφουν λιγότερο και λιγότερο κοινές (σε μια κλίμακα π.χ. από 1 σε 5). Για παράδειγμα:

```
find_simmilar_movies_5("Pirates Of The Caribbean", M).
M = "Pirates Of The Caribbean: On Stranger Tides"
M = "The Chronicles Of Narnia"
M = "Prince Of Persia: The Sands Of Time"
...
```
Για παράδειγμα, το παραπάνω ερώτημα θα επιστρέφει αρκετά κοινές σε περιεχόμενο ταινίες με την ταινια "Pirates Of The Caribbean". Θα υπάρχουν και αντίστοιχα ερωτήματα που θα βρίσκουν λιγότερο όμοιες ταινίες. Ο δείκτης ομοιότητας των ταινιών είναι αυθαίρετος και μπορείτε να τον ορίσετε εσείς όπως θέλετε, αρκεί να υπάρχει κάποια λογική σύνδεση με τα δεδομένα που περιέχονται στο αρχείο movies_metadata.csv. 

Συνεπώς η συνάρτηση που θα κάνει τις προτάσεις (recommendation) με είσοδο μια ταινία πρέπει να επιστρέφει-εκτυπώνει μια λίστα με τις προτεινόμενες ταινίες κατά φθίνουσα σειρά ομοιότητας. 


###3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις - Αξιολογήσεις του χρήστη


Σε αυτό το σημείο θα εργαστείτε με τα αρχεία ratings τα όποια περιέχουν αξιολογήσεις (από 1 μέχρι 5) για τις παραπάνω ταινίες. Το προηγούμενο σύστημα συστάσεων προτείνει στον χρήστη ταινίες αποκλειστικά με βάση την ομοιότητά τους. Σε αυτό το σημείο θα γίνει μια αναβάθμιση του συστήματος έτσι ώστε να παράγονται καλύτερες  συστάσεις οι όποιες θα λαμβάνουν υπόψιν και τις προτιμήσεις του χρήστη, οι όποιες θα εξάγονται από τις αξιολογήσεις που έχει κάνει μέχρι στιγμής. 

Η εκπαίδευση του recommender θα γίνεται ως εξής:

Για κάθε ταινία θα υπάρχει ένα score το όποιο αρχικά θε είναι ίσο με 0 και θα διαμορφώνεται από τις αξιολογήσεις κάθε user. Έτσι για έναν χρήστη με βάση τις αξιολογήσεις που υπάρχουν στο αρχείο train_ratings θα πρέπει:

1.	Για κάθε ταινία που έχει βαθμολογήσει να βρίσκονται οι κοινές ταινίες ανά κλίμακα και στο μέχρι τώρα σκορ κάθε παρόμοιας ταινίας θα προστίθεται ένα βάρος το όποιο θα μπορούσε να είναι το ποσοστό ομοιότητας της ταινίας (δηλαδή ένα βάρος για κάθε κλίμακα, αν δύο ταινίες μοιάζουν ενισχύουμε το βάρος που προσθέτουμε από το να μοιάζουν λιγότερο) επί τον βαθμό που έχει βάλει ο χρήστης για την αρχική ταινία (διαφορετικό είναι ο χρήστης να έχει βάλει 5/5 ή 1/5 σε μια ταινία από 3/5). 

2. Στη συνέχεια, ανάλογα με το σκορ που έχει σχηματιστεί για κάθε ταινία, θα επιλέγεται αν αυτή θα μπορούσε να είναι προτεινόμενη για τον χρήστη ή όχι και θα μετράμε το πόσο καλά τα πήγε το σύστήμά μας με βάση ορισμένες μετρικές.

Η λογική πίσω από την παραπάνω διαδικασία είναι ότι παρόμοιες ταινίες θα έχουν ανάλογο βαθμό. Για παράδειγμα αν ένας χρήστης έχει αξιολογήσει αρκετές  ταινίες οι όποιες είναι sci-fiction με 5/5 τότε μια ταίνια sci-fiction την όποια δεν έχει δεί λογικά θα του αρέσει και θα έπρεπε να την προτείνουμε.

Μετά την εκπαίδευση του συστήματος σας καλείστε να δοκιμάσετε τον recommender που κατασκευάσατε στην πράξη. Για τον σκοπό αυτό θα φορτώσετε το αρχείο test_ratings.csv όπου περιέχονται οι αξιολογήσεις του ίδιου χρήστη για άλλες ταινίες. Το σύστημά σας πρέπει να προβλέπει αν μια ταινία θα πρέπει να προταθεί στον χρήστη. Μια ταίνια θα έπρεπε να έχει προταθεί στον  χρήστη, αν έχει βαθμό μεγαλύτερο του 3. Συνεπώς για την επίβλεψη του συστήματός σας θα πρέπει για κάθε μια από τις ταινίες του αρχείου test_ratings.csv να επιστρέφεται 1 ή 0 αν η ταινία θα έπαιρνε βαθμό μεγαλύτερο του 3 ή όχι, δηλαδή αν θα έπρεπε να την είχαμε προτείνει ή όχι.

Στην συνέχεια, σε συνδιασμό με τις πραγματικές απαντήσεις του χρήστη θα αξιολογήσετε το σύστημά σας χρησιμοποιόντας τις μετρικές: precision, recall, f1 οι οποίες είναι οι πλέον γνωστές μετρικές και ευρέως χρησιμοποιούμενες τεχνικές για την επίβλεψη-μέτρηση απόδοσης ανάλογων συστημάτων.

1. Precision: Δείχνει πόσο ακριβές είναι το σύστημα. Υπολογίζει πόσα από τα στιγμιότυπα τα όποια προβλέψαμε ότι ανήκουν σε μια κλάση όντως ανήκουν σε αυτή. Η μετρική αυτή μας δίνει μια εικόνα σχετικά με τον αριθμό των ταινίων που προβλέψαμε ως προτεινόμενες ενώ δεν θα έπρεπε.

2. Recall: Υπολογίζει πόσα από τα στιγμιότυπα που ανήκουν σε μια κλάση (π.χ. προτεινόμενες ταινίες) προβλέφθηκαν σωστά.

3. F1: Είναι ένας μέσος μεταξύ των παραπάνω δυο μετρικών, έτσι ώστε να διατηρείται μια ισορροπία μεταξύ τους. Υπολογίζεται από την παρακάτω σχέση:

$$F_1=2\frac{Precision\times{Recall}}{Precision+Recall}$$

Οι παραπάνω συναρτήσεις παρέχονται από την βιβλιοθήκη scikit-learn.

Τέλος για την καλύτερη επίβλεψη του συστήματος σας μπορείτε να εκπαιδεύσετε  τον recommender σας με ένα υποσύνολο ταινίων από ελάχιστες, λίγες μέχρι και πολλές (π.χ. 3, 5, 10, 50, ...) για να μελετήσετε κατά πόσο σας βοηθούν οι επιπλέον αξιολογήσεις κάθε φορά (δηλαδή κατά πόσο βελτιώνονται οι παραπάνω μετρικές στο test set). 

Έτσι π.χ. μπορείτε να εντοπίσετε περιπτώσεις όπως για παράδειγμα ότι με έναν recommender μπορεί να μην επιτυγχάνετε πολύ υψηλό σκορ όσο με άλλους, αλλά το βέλτιστο σκόρ σας επιτυγχάνεται πολύ γρηγόρα π.χ. με μόνο 10 ταινίες αντί 100. Έτσι για παράδειγμα αν ο αλγόριθμος τα πηγαίνει πολύ καλά για τρεις ταινίες και στη συνέχεια το σκορ βελτιώνεται ελάχιστα τότε αυτός ενδείκνυται για ένα σύστημα συστάσεων για νέους χρήστες όπου δεδομένου λίγων ταινιών ο αλγόριθμος είναι σε θέση να το να προτείνει καλές συστάσεις. Ενώ αν η καλύτερη απόδοση του αλγορίθμου σας είναι βέλτιστη με περισσότερες ταινίες π.χ. 50 τότε αυτός ο αλγόριθμος συστάσεων ενδείκνυται για παλιούς χρήστες με πολλές αξιολόγησεις. 

---
## Κατασκευή Περιβάλλοντος Εργασίας
---
### Διάβασμα Αρχείων στο Colab (Μόνο για το Colab)
Αν η υλοποίηση γίνει στο google colab τότε μπορεί να χρησιμοποιηθεί το google drive ως file system. Για να γίνει Mount το google drive τρέχουμε τον παρακάτω κώδικα και κλικάρουμε στο link που θα μας εμφανιστεί. 

```
from google.colab import drive
drive.mount('/content/drive')
import os
os.listdir('/content/drive/My Drive')
```
Έπειτα στην σελίδα που άνοιξε επιλέγουμε το mail μας και στο επόμενο παράθυρο που θα μας ανοίξει πατάμε Να επιτρέπεται. Στην συνέχεια αντιγράφουμε τον κωδικό που θα μας βγάλει και τον κάνουμε paste στο Input που έχει ανοίξει στο colab. Έτσι πλέον αν έχουμε ανεβάσει ένα αρχείο στο google drive μπορούμε να το βρούμε στην θέση:

```
movies_filename = '/content/drive/My Drive/' + movies_metadata.csv
```
Μπορούμε πλέον κανονικά να δουλέψουμε φτιάχνοντας φακέλους ή αρχεία και γενικότερα κάνοντας οτιδήποτε θα κάναμε αν ήμασταν τοπικά. 

### Prolog μέσω Python

Το πακέτο που θα χρησιμοποιηθεί για την επικοινωνία Python και Prolog είναι το pyswip (https://pypi.org/project/pyswip/). Για να δουλέψει το Pyswip  χρειάζεται να υπάρχει το Swi-Prolog το όποιο αν δουλεύουμε τοπικά πρέπει να το εγκαταστήσουμε, ακολουθώντας αντίστοιχες οδηγίες στην σελίδα του εργαλείου. Για να γίνει του Swi-Prolog η εγκατάσταση στο Google Colab πρέπει να τρέξουμε τον παρακάτω κώδικα:

```
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog/stable
!sudo apt-get update
!sudo apt-get install swi-prolog
```
Σε κάποιο σημείο της εκτέλεσης εμφανίζεται ένα μήνυμα ότι πρέπει να πατήσουμε enter σε ένα input για να συνεχίσει η διαδικασία. Έπειτα από αυτό η εκτέλεση θα συνεχίσει χωρίς κάποιο πρόβλημα.

Τέλος πρέπει να εγκαταστήσουμε το pyswip (**όπου και να δουλεύουμε**) όπως παρακάτω:

```
!pip isntall pyswip
```


---
#Κώδικας για κατασκευή περιβάλλοντος εργασίας

##Μόνο για Google Colab

Κώδικας για να γίνει Mount to Google Drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

import os
os.listdir('/content/drive/My Drive')
path = '/content/drive/My Drive/'

Mounted at /content/drive


In [ ]:
%%capture
#install swi-prolog
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog
!sudo apt-get update
!sudo apt-get install swi-prolog
#install pyswip
!pip install pyswip


# **Μέρος 1: Μελέτη των Metadata, Δημιουργία κόσμου και των βασικών queries.**


In [ ]:
import pandas as pd
from pyswip import Prolog

In [ ]:
# Η βιβλιοθήκη pandas είναι χρήσιμη για την εργασία με τέτοια δεδομένα
import pandas as pd 
# Διάβασμα του αρχείου 'movie_metadata.csv' 
data = pd.read_csv(path + "movies_metadata.csv") 
#Στο csv υπαρχούν κελία με nan τιμές
#Στις θέσεις αυτές βάζουμε 'UNK' πράγμα που  κάνουμε με την παρακάτω συνάρτηση
data.fillna("UNK", inplace=True)
# Preview the first 5 lines of the loaded data 
data.head()

Unnamed: 0     budget  ...      actor_2_name       actor_3_name
0           0  237000000  ...  Sigourney Weaver       Stephen Lang
1           1  300000000  ...   Keira Knightley  Stellan Skarsgård
2           2  245000000  ...       Léa Seydoux      Ralph Fiennes
3           3  250000000  ...       Gary Oldman      Anne Hathaway
4           4  260000000  ...   Samantha Morton       Willem Dafoe

[5 rows x 27 columns]

In [ ]:
def clean_text(text):
  text = text.replace(u'\xa0', u'')
  text = text.replace(u"'", u'')
  return text

In [ ]:
#create World
#Ορίζουμε τον κόσμο μας
prolog = Prolog()

#Για κάθε row του πίνακα φτιάχνουμε τα κατηγορήματα που θέλουμε να αποθηκέυσουμε
#αρχικά σε μια λίστα με το όνομα literals
literals = []
movie_score = {}
for row in data.itertuples(index=True, name='Pandas'):
  movie_title = clean_text(getattr(row, 'movie_title'))
  if movie_title != 'UNK':

    for genre in getattr(row, 'genres').split("|"):
      if genre != 'UNK':
        literals.append("genre('"+ movie_title +"','"+ genre +"')")

    for director in getattr(row, 'director_name').split("|"):
      if director != 'UNK':
        literals.append("director('"+ movie_title +"','"+ clean_text(director) +"')")

    baw = 0
    for plotKeywords in getattr(row, 'plot_keywords').split("|"):
      if plotKeywords != 'UNK':
        if plotKeywords == "black and white":
          literals.append("colour('"+ movie_title +"','"+ "black_and_white" +"')")
          baw = 1
        else:
          literals.append("plotKeyWord('"+ movie_title +"','"+ clean_text(plotKeywords) +"')")
    if baw == 0:
      literals.append("colour('"+ movie_title +"','"+ "in_colour" +"')")
          

    for language in getattr(row, 'language').split("|"):
      if language != 'UNK':
        literals.append("language('"+ movie_title +"','"+ language +"')")

    for releaseyear in str(getattr(row, 'title_year')).split("|"):
      if releaseyear != 'UNK':
        literals.append("releaseYear('"+ movie_title +"','"+ releaseyear +"')")

    for duration in str(getattr(row, 'duration')).split("|"):
      if duration != 'UNK':
        literals.append("duration('"+ movie_title +"','"+ duration +"')")

    i = 0
    previous = -2
    for studio in getattr(row, 'production_companies').split("'"):
      i+=1
      if i-previous == 6:
        previous = i
        if studio != 'UNK':
          literals.append("production('"+ movie_title +"','"+ clean_text(studio) +"')")

    for country in getattr(row, 'country').split("|"):
      if country != 'UNK':
        literals.append("country('"+ movie_title +"','"+ clean_text(country) +"')")
    
    for actor1 in getattr(row, 'actor_1_name').split("|"):
      if actor1 != 'UNK':
        literals.append("actor('"+ movie_title +"','"+ clean_text(actor1) +"')")

    for actor2 in getattr(row, 'actor_2_name').split("|"):
      if actor2 != 'UNK':
        literals.append("actor('"+ movie_title +"','"+ clean_text(actor2) +"')")

    for actor3 in getattr(row, 'actor_3_name').split("|"):
      if actor3 != 'UNK':
        literals.append("actor('"+ movie_title +"','"+ clean_text(actor3) +"')")

#Η Prolog θέλει τα κατηγορήματά της με την σειρά 
literals.sort()
for literal in literals:
  prolog.assertz(literal)
  # print (literal +'.')

#Επίσης μπορούμε να κάνουμε consult ένα έτοιμο αρχείο στον κόσμο όπως παρακάτω
# prolog.consult(path + "db.pl")

####Σχολιασμός
Το movie dataset, αποτελείται από 26 attributes και 4801 entries. Τα κατηγορήματα που δημιουργούμε με βάση το τι θα μας φανεί χρήσιμο στη συνέχεια είναι: το genre(Movie, Genre), το director(Movie, Director), plotKeywords(Movie, PlotKeyword), colour(Movie, Colour), language(Movie, Language), releaseYear(Movie, ReleaseYear), studio(Movie, Studio), country(Movie, Country), actor(Movie, Actor).
Το αν η ταινία είναι έγχρωμη ή όχι το βρίσκουμε μέσα από τα plot_keywords, καθώς οι ασπρόμαυρες ταινίες έχουν ως plot_keyword το sting: ‘black and white’. Αν μια ταινία δεν το έχει τότε τη θεωρούμε έγχρωμη.
Για να βρούμε τα studio κάθε ταινίας, χωρίζουμε το attribute ‘production companies’ σε strings που χωρίζονται μεταξύ τους με τον χαρακτήρα “ ’ ” . Παρατηρούμε ότι τα ονόματα των studio βρίσκονται στις θέσεις 4, 10, 16, 22,... έτσι αρχικοποιούμε το prev στην τιμή -2, και όταν το τωρινό Index - prev = 6 προσθέτουμε το όνομα του studio στο κατηγόρημα και Prev=τωρινό index. 


####Μέρος 1β

In [ ]:
#1.-3. κοινό θέμα κλίμακα 5 - 1
prolog.assertz('''(similar_genre_5(X, Y):-
                                        genre(X,G1), genre(X,G2), genre(X,G3), genre(X,G4), genre(X,G5),
                                        genre(Y,G1), genre(Y,G2), genre(Y,G3), genre(Y,G4), genre(Y,G5),
                                        X \= Y, G1 \= G2, G1 \= G3, G1 \= G4, G1 \= G5, G2 \= G3, G2 \= G4, G2 \= G5, G3 \= G4, G3 \= G5, G4 \= G5 )''')

prolog.assertz('''(similar_genre_4(X, Y):-
                                        genre(X,G1), genre(X,G2), genre(X,G3), genre(X,G4), 
                                        genre(Y,G1), genre(Y,G2), genre(Y,G3), genre(Y,G4), 
                                        X \= Y, G1 \= G2, G1 \= G3, G1 \= G4, G2 \= G3, G2 \= G4, G3 \= G4)''')

prolog.assertz('''(similar_genre_3(X, Y):-
                                        genre(X,G1), genre(X,G2), genre(X,G3), 
                                        genre(Y,G1), genre(Y,G2), genre(Y,G3),
                                        X \= Y, G1 \= G2, G1 \= G3, G2 \= G3)''')

prolog.assertz('''(similar_genre_2(X, Y):-
                                        genre(X,G1), genre(X,G2),
                                        genre(Y,G1), genre(Y,G2),
                                        X \= Y, G1 \= G2)''')

prolog.assertz('(similar_genre_1(X, Y):- genre(X,G1), genre(Y,G1), X \= Y)')

In [ ]:
#4. κοινός σκηνοθέτης
prolog.assertz('(common_director(X, Y):- director(X,D), director(Y,D), X \= Y)')

In [ ]:
#5.-6. κοινή πλοκή κλίμακα 5 - 1
prolog.assertz('''(similar_plot_5(X, Y):-
                                        plotKeyWord(X,G1), plotKeyWord(X,G2), plotKeyWord(X,G3), plotKeyWord(X,G4), plotKeyWord(X,G5),
                                        plotKeyWord(Y,G1), plotKeyWord(Y,G2), plotKeyWord(Y,G3), plotKeyWord(Y,G4), plotKeyWord(Y,G5),
                                        X \= Y, G1 \= G2, G1 \= G3, G1 \= G4, G1 \= G5, G2 \= G3, G2 \= G4, G2 \= G5, G3 \= G4, G3 \= G5, G4 \= G5 )''')

prolog.assertz('''(similar_plot_4(X, Y):-
                                        plotKeyWord(X,G1), plotKeyWord(X,G2), plotKeyWord(X,G3), plotKeyWord(X,G4), 
                                        plotKeyWord(Y,G1), plotKeyWord(Y,G2), plotKeyWord(Y,G3), plotKeyWord(Y,G4), 
                                        X \= Y, G1 \= G2, G1 \= G3, G1 \= G4, G2 \= G3, G2 \= G4, G3 \= G4)''')

prolog.assertz('''(similar_plot_3(X, Y):-
                                        plotKeyWord(X,G1), plotKeyWord(X,G2), plotKeyWord(X,G3), 
                                        plotKeyWord(Y,G1), plotKeyWord(Y,G2), plotKeyWord(Y,G3),
                                        X \= Y, G1 \= G2, G1 \= G3, G2 \= G3)''')

prolog.assertz('''(similar_plot_2(X, Y):-
                                        plotKeyWord(X,G1), plotKeyWord(X,G2),
                                        plotKeyWord(Y,G1), plotKeyWord(Y,G2),
                                        X \= Y, G1 \= G2)''')

prolog.assertz('(similar_plot_1(X, Y):- plotKeyWord(X,G1), plotKeyWord(Y,G1), X \= Y)')

In [ ]:
#7.-9. κοινοί ηθοποιοί 3 - 1
prolog.assertz('''(common_actors_3(X, Y):-
                                        actor(X,A1), actor(X,A2), actor(X,A3), 
                                        actor(Y,A1), actor(Y,A2), actor(Y,A3),
                                        X \= Y, A1 \= A2, A1 \= A3, A2 \= A3)''')

prolog.assertz('''(common_actors_2(X, Y):-
                                        actor(X,A1), actor(X,A2),
                                        actor(Y,A1), actor(Y,A2),
                                        X \= Y, A1 \= A2)''')

prolog.assertz('(common_actors_1(X, Y):- actor(X,A1), actor(Y,A1), X \= Y)')

In [ ]:
#10. κοινή γλώσσα
prolog.assertz('(common_language(X, Y):- language(X,L), language(Y,L), X \= Y)')

In [ ]:
#11. έγχρωμες ή ασπρόμαυρες
prolog.assertz('(common_colour(X, Y):- colour(X,C), colour(Y,C), X \= Y)')

In [ ]:
#12. κοινό studio παραγωγής
prolog.assertz('(common_studio(X, Y):- production(X,S), production(Y,S), X \= Y)')

In [ ]:
#13. κοινή χώρα παραγωγής
prolog.assertz('(common_country(X, Y):- country(X,C), country(Y,C), X \= Y)')

In [ ]:
#14. Ίδια δεκαετία
prolog.assertz('''(common_decade(X, Y):- releaseYear(X,Y1), releaseYear(Y,Y2), 
                  X \= Y, atom_number(Y1, NY1), atom_number(Y2,NY2), IntY1 is integer(NY1), IntY2 is integer(NY2),
                  DY1 is IntY1//10, DY2 is IntY2//10, DY1==DY2)''')

In [ ]:
#ίδια διάρκεια
prolog.assertz('(common_duration(X, Y):- duration(X,T), duration(Y,T), X \= Y)')

####Σχολιασμός
Για τα genres και την πλοκή έχουμε φτιάξει κανόνες κλιμάκωσης από 1 μέχρι 5, ώστε να μπορούμε να έχουμε καλύτερη κλιμάκωση στο simple_recommender που θα δημιουργήσουμε στη συνέχεια.
Για τον κανόνα της ίδια δεκαετίας, διαιρούμε την χρονιά που κυκλοφόρησε η ταινία με το 10 (ακαίραια διαίρεση) ώστε να μας μείνουν μόνο τα 3 ψηφία που ορίζουν μια δεκαετία (πχ για το 2021 μένουν τα 202).

In [ ]:
# example
queries = ["similar_genre_5('","similar_genre_4('","similar_genre_3('","similar_genre_2('","similar_genre_1('", "common_director('",
      "similar_plot_5('","similar_plot_4('","similar_plot_3('","similar_plot_2('","similar_plot_1('","common_actors_3('","common_actors_2('",
      "common_actors_1('","common_language('", "common_colour('", "common_studio('","common_country('", "common_decade('", "common_duration('"]
def TEST(movie):
  for q in queries: 
    querry = prolog.query(q + movie +"',Y)")
    s = set()
    for soln in querry:
        s.add(soln["Y"])
    print(q.split("(")[0],list(set(s))[:40])
  return
  
print('Frankenweenie\n') 
TEST("Frankenweenie")
print('\nPirates of the Caribbean: The Curse of the Black Pearl\n')
TEST("Pirates of the Caribbean: The Curse of the Black Pearl")

Frankenweenie

similar_genre_5 []
similar_genre_4 []
similar_genre_3 ['Cloudy with a Chance of Meatballs 2', 'Happily NEver After', 'Big Hero 6', 'Hotel Transylvania', 'Free Birds', 'Ice Age: Continental Drift', 'Rio', 'VeggieTales: The Pirates Who Dont Do Anything', 'Despicable Me 2', 'A Bugs Life', 'Curious George', 'Megamind', 'The Ant Bully', 'Garfield: A Tail of Two Kitties', 'TMNT', 'Up', 'The Emperors New Groove', 'Rango', 'Alvin and the Chipmunks: The Squeakquel', 'Kung Fu Panda 3', 'The Boxtrolls', 'Dougs 1st Movie', 'Garfield', 'Shrek', 'Cats Dont Dance', 'Ice Age: The Meltdown', 'Foodfight!', 'The Smurfs', 'Antz', 'Monsters, Inc.', 'Hop', 'Toy Story 2', 'Surfs Up', 'Penguins of Madagascar', 'Aladdin', 'Legend of a Rabbit', 'The Nut Job', 'Monster House', 'Fantastic Mr. Fox', 'Home']
similar_genre_2 ['Are We There Yet?', 'Happily NEver After', 'Unaccompanied Minors', 'Curious George', 'Mr. Beans Holiday', 'My Girl', 'A Turtles Tale: Sammys Adventures', 'Back to the Future', '

In [ ]:
def get_num_of_genres(movie):
  q = prolog.query("genre('" + movie +"',Y)")
  s = set()
  for soln in q:
    m = soln['Y'] 
    if m not in s:
      s.add(soln['Y'])
  return len(s)

print(get_num_of_genres('Home Alone'))

2 10


# **2ο Μέρος: Recommendation System με βάση μόνο τα χαρακτηριστικά των ταινιών.**

Στο σημείο αυτό με βάση τους κανόνες που κατασκευάστηκαν στο Μέρος 1 θα κατασκευαστούν κατηγορήματα για την ομοιότητα ταινιών. Παρακάτω δίνεται ένα μικρό παράδειγμα ενός κανόνα και πώς αυτός θα μπορούσε να γραφτεί μέσω του Pyswip. Επίσης όπως αναφέρεται και σε σχόλιο παραπάνω θα μπορεί να γραφτεί και μια βάση δεδομένων με τους κανόνες και να γίνει απευθείας consult.

Στο παρακάτω παράδειγμα το 5 και το 4 εκφράζουν την ομοιότητα των ταινιών π.χ. οι ταινίες που παράγονται μέσω του find_similar_5 είναι πιο όμοιες από αυτές που παράγονται μέσω του find_similar_4. 


In [ ]:
prolog.assertz('''(find_sim_5(X, Y, G):- G ==5, similar_genre_5(X,Y), similar_plot_5(X,Y), common_director(X,Y), common_actors_1(X,Y), common_language(X,Y), common_colour(X,Y), (common_studio(X,Y); common_decade(X,Y)))''')
prolog.assertz('''(find_sim_5(X, Y, G):- G ==4, similar_genre_4(X,Y), similar_plot_5(X,Y), common_director(X,Y), common_actors_1(X,Y), common_language(X,Y), common_colour(X,Y), (common_studio(X,Y); common_decade(X,Y)))''')
prolog.assertz('''(find_sim_5(X, Y, G):- G ==3, similar_genre_3(X,Y), similar_plot_5(X,Y), common_director(X,Y), common_actors_1(X,Y), common_language(X,Y), common_colour(X,Y), (common_studio(X,Y); common_decade(X,Y)))''')

prolog.assertz('''(find_sim_4(X, Y, G):- G==4, similar_genre_4(X,Y), similar_plot_4(X,Y), common_language(X,Y), common_colour(X,Y))''')
prolog.assertz('''(find_sim_4(X, Y, G):- G==3, similar_genre_3(X,Y), similar_plot_4(X,Y), common_language(X,Y), common_colour(X,Y))''')
prolog.assertz('''(find_sim_4(X, Y, G):- G==2, similar_genre_2(X,Y), similar_plot_4(X,Y), common_language(X,Y), common_colour(X,Y))''')

prolog.assertz('''(find_sim_3(X, Y, G):- G==3, similar_genre_3(X,Y), similar_plot_3(X,Y), common_language(X,Y), common_colour(X,Y);
                                         G==3, similar_genre_3(X,Y), similar_plot_2(X,Y), common_language(X,Y), common_colour(X,Y))''')
prolog.assertz('''(find_sim_3(X, Y, G):- G==2, similar_genre_2(X,Y), similar_plot_3(X,Y), common_language(X,Y), common_colour(X,Y))''')

prolog.assertz('''(find_sim_2(X, Y):- similar_genre_1(X,Y), similar_plot_2(X,Y), common_language(X,Y), common_colour(X,Y);
                                      similar_genre_2(X,Y), similar_plot_1(X,Y), common_language(X,Y), common_colour(X,Y))''')

prolog.assertz('(find_sim_1(X, Y):- similar_genre_1(X,Y), similar_plot_1(X,Y))')

In [ ]:
def simple_recommender(movie):
    G = get_num_of_genres(movie)
    s5 = set()
    q = prolog.query("find_sim_5('" + movie +"',M, %d)" %G)
    for soln in q:
        m = soln['M']
        if m not in s5:
            s5.add(soln['M'])
    q.close()

    s4 = set()
    q = prolog.query("find_sim_4('" + movie +"',M, %d)" %G)
    for soln in q:
        m = soln['M']
        if m not in s4 and m not in s5:
            s4.add(soln['M'])
    q.close()

    s3 = set()
    q = prolog.query("find_sim_3('" + movie +"',M, %d)" %G)
    for soln in q:
        m = soln['M'] 
        if m not in s3 and m not in s4 and m not in s5:
            s3.add(soln['M'])
    q.close()

    s2 = set()
    q = prolog.query("find_sim_2('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s2 and m not in s3 and m not in s4 and m not in s5:
            s2.add(soln['M'])
    q.close()

    s1 = set()
    q = prolog.query("find_sim_1('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s1 and m not in s2 and m not in s3 and m not in s4 and m not in s5:
            s1.add(soln['M'])
    q.close()
    answers = set()
    for s in s5:
      answers.add((5, s))
    for s in s4:
      answers.add((4, s))
    for s in s3:
      answers.add((3, s))
    for s in s2:
      answers.add((2, s))
    for s in s1:
      answers.add((1, s))
    sorted_answers = sorted(answers, key=lambda tup: tup[0], reverse=True)
    return sorted_answers

In [ ]:
list(simple_recommender('Iron Man'))

[(4, 'Avengers: Age of Ultron'),
 (4, 'The Avengers'),
 (4, 'Iron Man 3'),
 (4, 'Iron Man 2'),
 (4, 'Ant-Man'),
 (4, 'Captain America: The Winter Soldier'),
 (3, 'The Wolverine'),
 (3, 'X-Men: Days of Future Past'),
 (3, 'X2'),
 (3, 'Superman Returns'),
 (3, 'Superman'),
 (3, 'X-Men Origins: Wolverine'),
 (3, 'Man of Steel'),
 (3, 'X-Men'),
 (3, 'Teenage Mutant Ninja Turtles'),
 (3, 'Superman III'),
 (3, 'The Shadow'),
 (3, 'X-Men: The Last Stand'),
 (3, 'Fantastic Four'),
 (3, 'Superman II'),
 (3, 'Mystery Men'),
 (3, 'Guardians of the Galaxy'),
 (2, 'The Fifth Element'),
 (2, 'Ghost Rider: Spirit of Vengeance'),
 (2, 'Hellboy II: The Golden Army'),
 (2, 'Brave'),
 (2, 'Monkeybone'),
 (2, 'The Phantom'),
 (2, 'The Sorcerers Apprentice'),
 (2, 'Green Lantern'),
 (2, 'Spider-Man'),
 (2, 'Spider-Man 3'),
 (2, 'Big Hero 6'),
 (2, 'Batman Begins'),
 (2, 'Lethal Weapon 3'),
 (2, 'March or Die'),
 (2, 'Scott Pilgrim vs. the World'),
 (2, 'The Dark Knight'),
 (2, 'Batman v Superman: Dawn of J

Δοκιμάζοντας διαφορετικούς κανόνες και παρατηρώντας τα κατηγορήματα κάθε ταινίας παρατηρήσαμε ότι υπάρχουν περιπτώσεις όπου η μη βέλτιστη καταγραφή των στοιχείων της (στο αρχικό cvc) οδηγεί σε λάθος συμπεράσματα. Για παράδειγμα, οι ταινίες Iron Man, Iron Man 2 και Iron Man 3 γνωρίζουμε ότι έχουν κοινό πρωταγωνιστή τον Robert Downey Jr., όμως καμία από τις 3 ταινίες δεν τον αναφέρει ως ηθοποιό της. Το Home Alone 2 έχει για Keywords τις εξής: christmas, holiday, new york city, new york, οι οποίες είναι λίγες και όχι ιδιαίτερα περιγραφικές της πλοκής της ταινίας με αποτέλεσμα το Home Alone 1 να μην θεωρείται πολύ παρόμοια ταινία ενώ στην πραγματικότητα είναι σχεδόν ίδιες.

# **3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις-Αξιολογήσεις του χρήστη-Εκπαίδευση και Πρόβλεψη**


Αρχικά μελετάμε τις αξιολογήσεις κάθε χρήστη για να καταλάβουμε την δομή και τις πληροφορίες κάθε αρχείου.




In [ ]:
from tqdm.notebook import tqdm
from sklearn.metrics import precision_score, recall_score, f1_score
import numpy as np
import random

rating_weights = {0: -1, 1: -0.5, 2:0, 3:0, 4:0.5, 5:1}
score_weights = {i:i + 1 for i in range(5)} # ανάλογα με τα επίπεδα ομοιότητας που έχουν οριστεί στην simple_recommender

def train_recommender(ratings, rating_weights, score_weights, number_of_movies = 10):
    """
    Στην συνάρτηση αυτή μπορούμε να ορίζουμε ποιο υποσύνολο των αξιολογήσεων θα χρησιμοποιήσουμε για το train μαζί με τα βάρη ομοιότητας και σκορ
    Σε συνδυασμό με τον αριθμό των ταινιών που θέλουμε να χρησιμοποιήσουμε σαν σύνολο δεδομένων π.χ. 10 από τις 100 ή 3 από τις 100 κ.ο.κ
    Αν θέλουμε να χρησιμοποιήσουμε όλες τις ταινίες σαν training set τότε ορίζουμε το number_of_movies = - 1
    """

    if number_of_movies > len(ratings):
        number_of_movies = len(ratings)


    if number_of_movies != -1:
        indexes = random.sample(range(len(ratings)), number_of_movies)
        ratings = ratings.iloc[indexes]

    movie_score = {}
    for row in tqdm(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        similar_movies = simple_recommender(movie)

        for (similarity, similar_movie) in similar_movies:
            if similar_movie not in movie_score:
                movie_score[similar_movie] = rating_weights[int(rating)] * score_weights[similarity - 1]
            else:
                movie_score[similar_movie] += rating_weights[int(rating)] * score_weights[similarity - 1] #το weight θα το ορίσετε ανα επίπεδο ομοιότητας οι πολύ όμοιες ταινίες θα έχουν μεγαλύτερο βάρος
    return movie_score


#αυτό είναι ένα παράδειγμα για το πως θα μπορούσε να υλοποιήθει η predict
#έχουμε ορίσει ότι μια ταινία θα έπρεπε να είναι προτεινόμενη αν είχε σκορ > 0
def predict_example(ratings, movie_score):
    real, pred = [], []
    for i, row in enumerate(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if movie in movie_score: #αν έχουμε σχηματίσει βαθμολογία για την ταινία αυτή
            pred.append(int(movie_score[movie] > 0)) #heuristic για το αν μια ταινία είναι προτεινόμενη
            real.append(int(rating > 3))# έτσι ορίζουμε ότι μια ταινία θα έπρεπε να είναι προτεινόμενη
            #η συνθήκη αυτή δεν μπορεί να αλλάξει
        else: #δεν μπορούμε να προτείνουμε κάτι για το όποιο δεν έχουμε σχηματίσει εικόνα
            pred.append(0)
            real.append(int(rating > 3))

    return real, pred

def get_metrics(real, pred):
    metrics = {}
    metrics["precision"] = precision_score(real, pred)
    metrics["recall"] = recall_score(real, pred)
    metrics["f1"] = f1_score(real, pred)
    return metrics

In [ ]:
train_ratings = pd.read_csv(path + "train_ratings.csv")
test_ratings = pd.read_csv(path + "test_ratings.csv")

In [ ]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, -1)
real, pred = predict_example(test_ratings, movie_score)
print(get_metrics(real, pred))

0it [00:00, ?it/s]

{'precision': 0.5180722891566265, 'recall': 0.5972222222222222, 'f1': 0.5548387096774194}


####Σχολιασμός
Παρατηρούμε ότι έχουμε precision 52% και recall 60% ποσοστά τα οποία είναι σχετικά ικανοποιητικά. Δοκιμάζοντας διαφορετικές παραμέτρους ως βάρη καταλλήξαμε στα παραπάνω βάρη. Όπως αναφέραμε παραπάνω τα αρχικά δεδομένα οδηγούν σε λανθασμένες ή ελλειπής προβλέψεις με αποτέλεσμα να μην έχουμε πολύ υψηλό percision και recall λόγω του περιεχομένου του dataset.